File: lib/glg/SimpleViewer/MultiThreaded/C/demo_datafeed.c
#include "GlgViewer.h"

#define SLEEP_TIME  25    /* millisec */

#define RELATIVE_TO_NEW_RANGE( low, high, rel_value ) \
   ( (low) + ((high) - (low)) * rel_value )

void ProcessDemoData( void * sender_socket );
GlgBoolean GetGPSData( GPSData * data );
GlgBoolean GetTelemetryData( TelemetryData * data );
double GetDemoValue( double low, double high );
void GetDemoPosition( GlgPoint * latlon, double lon_min, double lon_max,
                      double lat_min, double lat_max,
                      double alt_low, double alt_high );

/*--------------------------------------------------------------------
| Get data from an application data source and send them to GUI via
| the provided sender socket.
*/
void ProcessDemoData( void * sender_socket )
{
   GPSData gps_data;
   TelemetryData telemetry_data;

   /* In the demo, use separate methods (GetGPSData and GetTelemetryData)
      to generate demo data for GPS and telemetry. In an application,
      a single method can be used to receive all data from either a single
      source/socket, or a multiple sources.
   */

   if( GetGPSData( &gps_data ) )
     SendDataToGUIThread( sender_socket, (BaseData*) &gps_data );

   if( GetTelemetryData( &telemetry_data ) )
     SendDataToGUIThread( sender_socket, (BaseData*) &telemetry_data );

   /* For a demo, put the thread to sleep after generateing data. 
      In an application, a single method that receives all data may simply 
      block if there are no incoming data.
   */
   SleepMS( SLEEP_TIME );
}

#define LOW 0
#define HIGH 100
#define MIN_LAT -90.
#define MAX_LAT 90.
#define MIN_LON -180.
#define MAX_LON +180.
#define ALT_LOW 0.
#define ALT_HIGH 10000.

/*--------------------------------------------------------------------
| Receive GPS data (using random values for demo) and store in the
| provided buffer. If no GPS data available, return False.
*/
GlgBoolean GetGPSData( GPSData * data )
{
   GlgPoint latlon;

   data->base.type = GPS;
   data->base.is_valid = GlgTrue;
   data->base.time_stamp = GetCurrTime();

   GetDemoPosition( &latlon, MIN_LON, MAX_LON,
                    MIN_LAT, MAX_LAT, ALT_LOW, ALT_HIGH );

   data->lat = latlon.y; 
   data->lon = latlon.x; 
   data->altitude = latlon.z;
   data->speed = GetDemoValue( LOW, HIGH ); 
   data->pitch = GetDemoValue( -30., 30. );
   data->roll = GetDemoValue( -30., 30. );
   data->yaw = GetDemoValue( -30., 30. );

   return True;
}

/*--------------------------------------------------------------------
| Receive TELEMETRY data (using random values for demo) and store in 
| the provided buffer. If no TELEMETRY data available, return False.
*/
GlgBoolean GetTelemetryData( TelemetryData * data )
{
   data->base.type = TELEMETRY;
   data->base.is_valid = GlgTrue;
   data->base.time_stamp = GetCurrTime();

   data->power = GetDemoValue( LOW, HIGH );  
   data->voltage = GetDemoValue( LOW, HIGH );  
   data->current = GetDemoValue( LOW, HIGH );  
   data->temperature = GetDemoValue( LOW, HIGH );  
   data->pressure = GetDemoValue( LOW, HIGH );  
   data->state_health = GetDemoValue( 0., 3. );

   return True;
}

/*----------------------------------------------------------------------*/
double GetDemoValue( double low, double high )
{
#define PERIOD    5000
   static GlgLong counter = 0;
   double 
     value, 
     angle; 

   ++counter;
   angle = ( counter % PERIOD ) / (double) PERIOD;
   value = 0.5 + sin( 2. * M_PI * angle ) / 2.;      
   
   value = RELATIVE_TO_NEW_RANGE( low, high, value );
   
   return value;
}

/*----------------------------------------------------------------------*/
void GetDemoPosition( GlgPoint * latlon, 
                      double lon_min, double lon_max,
                      double lat_min, double lat_max,
                      double alt_low, double alt_high )
{
#define LAT_PERIOD     1000
#define LON_PERIOD     4000
#define OFFSET1     0
#define OFFSET2     0

   static long counter = 0;
   double
     angle1, angle2,
     lat, lon;

   ++counter;
   
   angle1 = ( counter % LAT_PERIOD ) / (double) LAT_PERIOD;
   angle2 = ( counter % LON_PERIOD ) / (double) LON_PERIOD;

   lat = 0.5 + sin( 2. * M_PI * angle1 + OFFSET1 ) / 2. * 0.9;
   lon = 0.5 + sin( 2. * M_PI * angle2 + OFFSET2 ) / 2. * 0.9;

   latlon->x = 
     RELATIVE_TO_NEW_RANGE( lon_min, lon_max, lon );

   latlon->y = 
     RELATIVE_TO_NEW_RANGE( lat_min, lat_max, lat );

   /* Get a simulated altitude value. */
   latlon->z = GetDemoValue( alt_low, alt_high );
}

/*--------------------------------------------------------------------
| Simulte the "Write" opearion. For demo purposes, set the value
| of the specified tag source in the drawing. 
*/
GlgBoolean DemoWriteDValue( char * tag_source, double d_value )
{
   extern GlgObject Viewport;

   return GlgSetDTag( Viewport, tag_source, d_value, False );  
}

File: lib/glg/SimpleViewer/MultiThreaded/Java/DemoDataFeed.java
import com.genlogic.*;

//////////////////////////////////////////////////////////////////////////
// DemoDataFeed provides simulated data for demo, as well as for testing 
// with no LiveDataFeed.
// In an application, data will be coming from LiveDataFeed.
//////////////////////////////////////////////////////////////////////////

public class DemoDataFeed extends DataFeedBase
{
   // Low/High range for simulated data. 
   double HIGH = 100.;
   double LOW = 0.;
   double MIN_LAT = -90.;
   double MAX_LAT = 90.;
   double MIN_LON = -180.;
   double MAX_LON = +180.;
   double ALT_LOW = 0.;
   double ALT_HIGH = 10000.;
   
   // Used to generate simulated data.
   long counter1 = 0;
   long counter2 = 0;

   // Sleep time in milliseconds.
   static final int SLEEP_MS = 2; 

   /////////////////////////////////////////////////////////////////////// 
   public DemoDataFeed( GlgViewer viewer ) 
   {
      super( viewer );
   }

   /////////////////////////////////////////////////////////////////////// 
   // Invoked asynchronously from a thread. Accumulates received data 
   // in AccumulatedData vector. 
   // 
   // Thread synchronization is used inside DataFeedBase.StoreRawData() 
   // to ensure synchronization for AccumulatedData object.
   /////////////////////////////////////////////////////////////////////// 
   @Override
   void ProcessData()
   {
      // Process GPS data (if any) and store them in AccumulatedData.
      GPSData gps_data = GetGPSData();
      if( gps_data != null )
        StoreRawData( gps_data );
      
      // Process TELEMETRY data (if any) and store them in AccumulatedData.
      TelemetryData telem_data = GetTelemetryData();
      if( telem_data != null )
        StoreRawData( telem_data );
      
      /* For a demo, put the data thread to sleep after processing data. 
         In an application, a method that receives data may simply 
         block if there are no incoming data.
      */
       try 
       {
          DataThread.sleep( SLEEP_MS );
       } 
       catch (InterruptedException e) 
       {
          System.out.println( "DataThread sleep interrupted." );
       }
   }
   
   ///////////////////////////////////////////////////////////////////////
   // Receive GPS data (using random values for demo).
   // If no GPS data available, return null.
   ///////////////////////////////////////////////////////////////////////
   GPSData GetGPSData()
   {
      GPSData data = new GPSData();
      
      GlgPoint latlon = 
        GetDemoPosition( MIN_LON, MAX_LON, MIN_LAT, MAX_LAT, 
                         ALT_LOW, ALT_HIGH );
      
      data.lat = latlon.y; 
      data.lon = latlon.x; 
      data.altitude = latlon.z;
      data.speed = GetDemoValue( LOW, HIGH ); 
      data.pitch = GetDemoValue( -30., 30. );
      data.roll = GetDemoValue( -30., 30. );
      data.yaw = GetDemoValue( -30., 30. );
      
      data.is_valid = true;
      data.time_stamp = GetCurrTime();
      
      return data;
   }

   ///////////////////////////////////////////////////////////////////////
   // Receive TELEMETRY data (using random values for demo).
   // If no TELEMETRY data available, return NULL.
   ///////////////////////////////////////////////////////////////////////
   TelemetryData GetTelemetryData()
   {
      TelemetryData data = new TelemetryData();
      
      data.power = GetDemoValue( LOW, HIGH );  
      data.voltage = GetDemoValue( LOW, HIGH );  
      data.current = GetDemoValue( LOW, HIGH );  
      data.temperature = GetDemoValue( LOW, HIGH );  
      data.pressure = GetDemoValue( LOW, HIGH );  
      data.state_health = GetDemoValue( 0., 3. );
      
      data.is_valid = true;
      data.time_stamp = GetCurrTime();
      
      return data;
   }
   
   /////////////////////////////////////////////////////////////////////// 
   @Override
   boolean WriteDValue( String tag_source, double d_value )
   {
      /* In the demo mode, set the tag in the drawing to simulate round trip
         after the write operation.
      */
      return Viewer.SetDTag( tag_source, d_value, false );
   }

   /////////////////////////////////////////////////////////////////////// 
   @Override
   boolean WriteSValue( String tag_source, String s_value )
   {
      /* In the demo mode, set the tag in the drawing to simulate round trip
         after the write operation.
      */
      return Viewer.SetSTag( tag_source, s_value, false );
   }
   
   /////////////////////////////////////////////////////////////////////// 
   double GetDemoValue( double low, double high )
   {
      int PERIOD = 5000;
      double 
        value, 
        angle; 

      ++counter1;
      angle = ( counter1 % PERIOD ) / (double) PERIOD;
      value = 0.5 + Math.sin( 2. * Math.PI * angle ) / 2.;      
      
      return ( value * ( high - low ) + low );
   }
   
   /////////////////////////////////////////////////////////////////////// 
   GlgPoint GetDemoPosition( double lon_min, double lon_max,
                                    double lat_min, double lat_max,
                                    double alt_low, double alt_high )
   {
      int LAT_PERIOD = 1000;
      int LON_PERIOD = 4000;
      int OFFSET1 = 0;
      int OFFSET2 = 0;

      double
        angle1, angle2,
        lat, lon;
      
      ++counter2;
   
      angle1 = ( counter2 % LAT_PERIOD ) / (double) LAT_PERIOD;
      angle2 = ( counter2 % LON_PERIOD ) / (double) LON_PERIOD;
      
      lat = 0.5 + Math.sin( 2. * Math.PI * angle1 + OFFSET1 ) / 2. * 0.9;
      lon = 0.5 + Math.sin( 2. * Math.PI * angle2 + OFFSET2 ) / 2. * 0.9;
      
      GlgPoint latlon = new GlgPoint();
      latlon.x = lon * ( lon_max - lon_min ) + lon_min;
      latlon.y = lat * ( lat_max - lat_min ) + lat_min;
            
      // Get a simulated altitude value.
      latlon.z = GetDemoValue( alt_low, alt_high );

      return latlon;
   }
}
   
   
